pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,459 @@
+        ensure!(!Zero::is_zero(&shares.0), Error::<T>::StakeMustBeNonZero);
+
+        // We create the new joining shares. It returns the actual amount of stake those shares
+        // represents (due to rounding).
+        let stake = pools::Joining::<T>::add_shares(&candidate, &delegator, shares.clone())?;
+
+        // We hold the funds of the delegator and register its stake into the candidate stake.
+        T::Currency::hold(&T::CurrencyHoldReason::get(), &delegator, stake.0)?;
+        pools::Joining::<T>::increase_hold(&candidate, &delegator, &stake)?;
+        Candidates::<T>::add_total_stake(&candidate, &stake)?;
+
+        // We create/mutate a request for joining.
+        let now = frame_system::Pallet::<T>::block_number();
+        let operation_key = match pool {
+            TargetPool::AutoCompounding => PendingOperationKey::JoiningAutoCompounding {
+                candidate: candidate.clone(),
+                at_block: now,
+            },
+            TargetPool::ManualRewards => PendingOperationKey::JoiningManualRewards {
+                candidate: candidate.clone(),
+                at_block: now,
+            },
+        };
+
+        // We store/mutate the operation in storage.
+        let operation = PendingOperations::<T>::get(&delegator, &operation_key);
+        let operation = operation
+            .err_add(&shares.0)
+            .map_err(|_| Error::<T>::MathOverflow)?;
+        PendingOperations::<T>::set(&delegator, &operation_key, operation);
+
+        Pallet::<T>::deposit_event(Event::<T>::RequestedDelegate {
+            candidate,
+            delegator,
+            towards: pool,
+            pending: stake.0,
+        });
+
+        Ok(().into())
+    }
+
+    pub fn request_undelegate(
+        origin: OriginFor<T>,
+        candidate: Candidate<T>,
+        pool: TargetPool,
+        amount: SharesOrStake<T::Balance>,
+    ) -> DispatchResultWithPostInfo {
+        let delegator = ensure_signed(origin)?;
+
+        // Converts amount to shares of the correct pool
+        let shares = match (amount, pool) {
+            (SharesOrStake::Shares(s), _) => s,
+            (SharesOrStake::Stake(s), TargetPool::AutoCompounding) => {
+                pools::AutoCompounding::<T>::stake_to_shares(&candidate, Stake(s))?.0
+            }
+            (SharesOrStake::Stake(s), TargetPool::ManualRewards) => {
+                pools::ManualRewards::<T>::stake_to_shares(&candidate, Stake(s))?.0
+            }
+        };
+
+        // Destroy shares
+        let removed_stake = match pool {
+            TargetPool::AutoCompounding => {
+                let stake = pools::AutoCompounding::<T>::sub_shares(
+                    &candidate,
+                    &delegator,
+                    Shares(shares),
+                )?;
+
+                if stake.0 > pools::AutoCompounding::<T>::hold(&candidate, &delegator).0 {
+                    Self::rebalance_hold(
```

Comment:
this basically means that the hold was not well balanced and as such we need to rebalance it before we decrease it correct?

Reply:
Exactly. Delegator "received" auto rewards but those are in the staking account without a rebalance. If the delegator wants to unstake all its stake + those rewards there is not enough held for that.

Note that this should only be necessary for auto compounding, since manual rewards are .. not compounding :p So this kind of "positive imbalance" should not occur (only negative on with slashing).